bigquery/src/main/java/org/apache/iceberg/gcp/bigquery/BigQueryMetaStoreClientImpl.java

```diff
@@ -0,0 +1,620 @@
+              DEFAULT_RETRY_CONFIG);
+    } catch (BigQueryRetryHelper.BigQueryRetryHelperException e) {
+      handleBigQueryRetryException(e);
+    }
+    return response;
+  }
+
+  @SuppressWarnings("FormatStringAnnotation")
+  private Dataset create(Dataset dataset) {
+    try {
+      HttpResponse response =
+          client
+              .datasets()
+              .insert(dataset.getDatasetReference().getProjectId(), dataset)
+              .executeUnparsed();
+      return convertExceptionIfUnsuccessful(response).parseAs(Dataset.class);
+    } catch (IOException e) {
+      throw new RuntimeIOException(e);
+    } catch (AlreadyExistsException e) {
+      throw new AlreadyExistsException("Namespace already exists: " + dataset.getId());
+    }
+  }
+
+  @Override
+  @SuppressWarnings("FormatStringAnnotation")
+  public Dataset getDataset(DatasetReference datasetReference) {
+    try {
+      HttpResponse response =
+          client
+              .datasets()
+              .get(datasetReference.getProjectId(), datasetReference.getDatasetId())
+              .executeUnparsed();
+      if (response.getStatusCode() == HttpStatusCodes.STATUS_CODE_NOT_FOUND) {
+        throw new NoSuchNamespaceException(
+            "Namespace does not exist: " + datasetReference.getDatasetId());
+      }
+      return convertExceptionIfUnsuccessful(response).parseAs(Dataset.class);
+    } catch (IOException e) {
+      throw new RuntimeIOException(e);
+    }
+  }
+
+  @Override
+  @SuppressWarnings("FormatStringAnnotation")
+  public void deleteDataset(DatasetReference datasetReference) {
+    try {
+      HttpResponse response =
+          client
+              .datasets()
+              .delete(datasetReference.getProjectId(), datasetReference.getDatasetId())
+              .executeUnparsed();
+      if (response.getStatusCode() == HttpStatusCodes.STATUS_CODE_NOT_FOUND) {
+        throw new NoSuchNamespaceException(
+            "Namespace does not exist: " + datasetReference.getDatasetId());
+      }
+      convertExceptionIfUnsuccessful(response);
+    } catch (IOException e) {
+      throw new RuntimeIOException(e);
+    }
+  }
+
+  @Override
+  public Dataset setDatasetParameters(
+      DatasetReference datasetReference, Map<String, String> parameters) {
+    Dataset dataset = getDataset(datasetReference);
+    ExternalCatalogDatasetOptions externalCatalogDatasetOptions =
+        dataset.getExternalCatalogDatasetOptions() == null
+            ? new ExternalCatalogDatasetOptions()
+            : dataset.getExternalCatalogDatasetOptions();
+    Map<String, String> finalParameters =
+        externalCatalogDatasetOptions.getParameters() == null
+            ? Maps.newHashMap()
+            : externalCatalogDatasetOptions.getParameters();
+    finalParameters.putAll(parameters);
+
+    dataset.setExternalCatalogDatasetOptions(
+        externalCatalogDatasetOptions.setParameters(finalParameters));
+
+    return updateDataset(dataset);
```

Comment:
What happens if the dataset is changed between L223-237 concurrently? Same for removeDatasetParameters.

Reply:
BigQuery uses optimistic locking aka ETag. If it is changed update will be failed. Do you see any issue for that ? @ebyhr